### PR TITLE
Numpy . add numpy op logspace

### PR DESCRIPTION
**Pull request by gyshi**
## Description ##
(Brief description on what this PR is about)

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/operator/tensor/init_op.h [hunk 1]
```diff
@@ -644,6 +644,75 @@ inline bool RangeLikeShape(const nnvm::NodeAttrs& attrs,
...
+            MXNET_ADD_ALL_TYPES
+        .describe("Target data type.");
+  }
+};
+
+struct logspace_fwd {
+  template<typename DType>
+  MSHADOW_XINLINE static void Map(index_t i, double start, double stop, double base,
+                                  double step, int req, DType* out) {
+    KERNEL_ASSIGN(out[i], req, static_cast<DType>(math::pow(base, static_cast<double>(start + step * i))));  // NOLINT(*)
```
Comment (haojin2) [inline src/operator/tensor/init_op.h]: Don't use `NOLINT`, just do line wrapping:
```c++
KERNEL_ASSIGN(out[i], req,
              static_cast<DType>(math::pow(base, static_cast<double>(start + step * i))));
```

File: src/operator/tensor/init_op.h [hunk 2]
```diff
@@ -644,6 +644,75 @@ inline bool RangeLikeShape(const nnvm::NodeAttrs& attrs,
...
+  const LogspaceParam& param = nnvm::get<LogspaceParam>(attrs.parsed);
+  MSHADOW_TYPE_SWITCH(outputs[0].type_flag_, DType, {
+    int step_num = param.endpoint ? param.num - 1 : param.num;
+    double step = step_num > 0 ? (param.stop - param.start) / step_num : 0.0f;
+    Kernel<logspace_fwd, xpu>::Launch(s, outputs[0].Size(), param.start, param.stop, param.base,
+                                      step, req[0], outputs[0].dptr<DType>());
+  });
+}
+
+inline bool LogspaceShape(const nnvm::NodeAttrs& attrs,
```
Comment (haojin2) [inline src/operator/tensor/init_op.h]: put this function in `.cc` file.

File: src/operator/tensor/init_op.h [hunk 3]
```diff
@@ -644,6 +644,75 @@ inline bool RangeLikeShape(const nnvm::NodeAttrs& attrs,
   return true;
 }
 
+struct LogspaceParam : public dmlc::Parameter<LogspaceParam> {
```
Comment (haojin2) [inline src/operator/tensor/init_op.h]: Create a `np_init_op.h` under `src/operator/numpy` folder and put all these code there.

File: src/operator/tensor/init_op.h (hunk 2 above)
Comment (gyshi) [reply src/operator/tensor/init_op.h]: i think i should put it in np_init_op.h file.

**Comment by gyshi** [comment]
@haojin2  @reminisce  thx for reviewing

File: tests/python/unittest/test_numpy_op.py [hunk 4]
```diff
@@ -1951,13 +1951,77 @@ def get_grad_right(a1, a2):
...
+                        np_ret = _np.logspace(*config, endpoint=endpoint, base=base, dtype=dtype)
+                    else:
+                        mx_ret = np.logspace(config, endpoint=endpoint, base=base, dtype=dtype)
+                        np_ret = _np.logspace(config, endpoint=endpoint, base=base, dtype=dtype)
+                    assert_almost_equal(mx_ret.asnumpy(), np_ret, atol=1e-3, rtol=1e-5)
+    # check for exception input
+    for config in exception_configs:
+        assertRaises(MXNetError, np.logspace, *config)
+
+    @use_np
```
Comment (reminisce) [inline tests/python/unittest/test_numpy_op.py]: This decorator should not be necessary.

File: src/operator/numpy/np_init_op.h [hunk 5]
```diff
@@ -101,6 +102,64 @@ void IndicesCompute(const nnvm::NodeAttrs& attrs,
...
+
+template<typename xpu>
+void LogspaceCompute(const nnvm::NodeAttrs& attrs,
+                     const OpContext& ctx,
+                     const std::vector<TBlob>& inputs,
+                     const std::vector<OpReqType>& req,
+                     const std::vector<TBlob>& outputs) {
+  using namespace mxnet_op;
+  Stream<xpu> *s = ctx.get_stream<xpu>();
+  const LogspaceParam& param = nnvm::get<LogspaceParam>(attrs.parsed);
```
Comment (reminisce) [inline src/operator/numpy/np_init_op.h]: Please avoid kernel launch when `outputs[0].Size()` is 0.

File: src/operator/numpy/np_init_op.h [hunk 6]
```diff
@@ -101,6 +102,64 @@ void IndicesCompute(const nnvm::NodeAttrs& attrs,
...
+  MSHADOW_TYPE_SWITCH(outputs[0].type_flag_, DType, {
+      int step_num = param.endpoint ? param.num - 1 : param.num;
+      double step = step_num > 0 ? (param.stop - param.start) / step_num : 0.0f;
+      Kernel<logspace_fwd, xpu>::Launch(s, outputs[0].Size(), param.start, param.stop, param.base,
+      step, req[0], outputs[0].dptr<DType>());
```
Comment (reminisce) [inline src/operator/numpy/np_init_op.h]: Indent by 4 spaces.

File: tests/python/unittest/test_numpy_op.py [hunk 7]
```diff
@@ -1951,13 +1951,77 @@ def get_grad_right(a1, a2):
...
+            return x + F.np.logspace(self._start, self._stop, self._num,
+                                     self._endpoint, self._base, self._dtype)
+
+    for dtype in dtypes:
+        x = np.zeros(shape=(), dtype=dtype)
+        for config in configs:
+            for hybridize in [False, True]:
+                for endpoint in [False, True]:
+                    for base in base_configs:
+                        if isinstance(config, tuple):
```
Comment (reminisce) [inline tests/python/unittest/test_numpy_op.py]: Seems all configs are `tuple`s in `configs`. What's the point of checking this here?

**Pull request merged**